Namespace File created

Smoren opened the pull request:
Hi @markrogoyski,

It is a draft of PR now.

Tell me, am I moving in the right direction? If yes, I will continue.

Summary:

Namespace `File` created with methods:
* `readByLine()`;
* `readCsv()`;
* `write()`.

Smoren committed:
Namespace `File` created with methods:
* `readByLine()`;
* `readCsv()`;
* `write()`.

Smoren committed:
little fixes

Smoren committed:
Fixes for codesniffer and phpstan. Workflow succeed.

Smoren committed:
`File` namespace: throw `RuntimeException` on errors.

markrogoyski commented:
Hi @Smoren,

Thanks for starting this discussion.

`File::readLines` and `File::readCsv` make sense as is to be used in loops, providing a simpler interface which doesn't have to deal with IO return values and loop-variable assignment.

The `File::write` I don't think fits with the IterTools loop construct pattern. I think the `write` method makes more sense as a Stream terminal operator. For example:
```php
Stream::ofFile($fh)
    ->filter( ... )
    ->map( ... )
    ->write('path/to/newFile.txt');
```

In a similar thought, the read/readCsv functions seem suitable for Stream source functions: `Stream::ofFile(...)` and `Stream::ofCsvFile`.

Mark

markrogoyski commented on src/File.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @see fgets()
+     */
+    public static function readByLine($fileResource, ?int $length = null): \Generator
```
What do you think about getting rid of the `$length` parameter to simplify this. I'm pretty sure 99% of users will not use it.

Smoren replied:
Agree, done.

markrogoyski commented on src/File.php:
```diff
@@ -0,0 +1,101 @@
+     * @see fgetcsv()
+     */
+    public static function readCsv(
+        $fileResource,
+        ?int $length = null,
+        string $separator = ",",
+        string $enclosure = "\"",
+        string $escape = "\\"
```
What do you think about rearranging the parameters to put the more-likely-to-be-used parameters first:
```php
readCsv(resource $file, string $separator = ',', string $enclosure = '"', string $escape = '\') { ... }
```
And how about removing length?

Smoren replied:
Agree. Done.

Smoren committed:
`File` namespace: applying initial suggestions.

Smoren commented:
Hi @markrogoyski,

Thank yo for the feedback!

This is a great ideas to use writing to a file as a stream terminal operation and reading file resources as stream source operations.

I'll implement this ideas.

Smoren committed:
`File` namespace: `ReadCsvTest` added.

Smoren committed:
Codesniffer fix

Smoren committed:
PHPdoc added for `FileFixture`.

Smoren committed:
`File` namespace: bug fixed, test added.

Smoren committed:
Error test commented

markrogoyski commented on tests/Fixture/FileFixture.php:
```diff
@@ -0,0 +1,28 @@
+     */
+    public static function createFromString(string $string)
+    {
+        $file = tmpfile();
+        fwrite($file, $string);
+        fseek($file, 0);
```
Suggestion: Take a look at this project: https://github.com/bovigo/vfsStream. I haven't looked too deeply at it, but it might allow you to write unit tests involving files without an infrastructure dependency on an actual file system.

Smoren replied:
Thank you! I'll try.

Smoren replied:
Hi @markrogoyski, [I've used](https://github.com/markrogoyski/itertools-php/pull/19/commits/f2f34f7eedecb7d27d6fb982cf4358e8d930fd60) the package you recommended. Now `FileFixture` do not use real filesystem.

markrogoyski replied:
Nice. Thanks for investigating that alternative as a way to interact with files without involving the actual file system. I'm glad it worked out. It will make for more reliable unit testing.

Smoren committed:
`File` namespace refactored, tests cases added.

Smoren committed:
`File` methods: checking resource is valid in reading process; test cases added.

Smoren committed:
`File` namespace tests: `vfsStream` used.

markrogoyski commented on tests/File/ReadCsvTest.php:
```diff
@@ -0,0 +1,364 @@
+                [['1', '2', '3']],
+            ],
+            [
+                [
+                    '1,2,3',
+                    '4,5,6',
+                ],
+                [",", "\"", "\\"],
+                [
+                    ['1', '2', '3'],
+                    ['4', '5', '6'],
```
Suggestion: For unit tests, we want to try to avoid any unnecessary logic via code, as we want to test as straightforwardly as possible our code without introducing other avenues for mistakes/bugs to slip in via code our tests. Therefore, consider possibly using [Nowdoc](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.nowdoc) or Heredoc syntax to create multi-line strings for processing, rather than manipulating nested array structures to create a multiline file. Now/Heredoc is kind of verbose and often results in code that breaks indentation, but I think it is more important to have tests that have simple setups (Given sections) to ensure the test is easy to understand there is no question what it is doing.

Smoren replied:
Hi @markrogoyski,

Do you think it's really necessary in this case? To be honest, I don't like this syntax precisely because it spoils the code design. It seems to me that there is quite a bit of logic in the construction of the fixture. But if you insist, I will of course rewrite using this syntax.

markrogoyski replied:
Yeah, I agree it is ugly syntax, especially when it breaks indentation. No need to rewrite any existing tests.

Smoren replied:
Thank you Mark!

Smoren committed:
Method `File::checkIsResourceValid()` became private.

Smoren committed:
`Stream` namespace: new methods are implemented (not tested yet):
- `ofFileLines()`,
- `ofCsvFile()`,
- `toFileLines()`,
- `toCsvFile()`.

Smoren committed:
Tests were added for `Stream` methods: `ofFileLines()`, `ofCsvFile()`; root namespace added for function calls in helpers.

markrogoyski commented on src/Stream.php:
```diff
@@ -676,6 +713,55 @@ public function toArray(): array
+     * @throws FileException
+     */
+    public function toCsvFile(
+        string $path,
```
Question: The read file methods are taking a `resource` handle as input. For consistency, wouldn't it make sense to also take in a resource handle for the write methods as well? It also avoids dealing with the file system issues of opening a file handle and having it fail.

Smoren replied:
I thought about such a solution, but I was confused that the user would have to open and close file resources on his own.

I think that from the point of view of the client, this is quite inconvenient.

Smoren replied:
BTW, we can add a new method `toResource()` that will take a resource as input.

markrogoyski replied:
Opening and passing around a file handle or database handle or curl handle, etc. are common practices. It is also consistent with the other methods. I recommend taking in a file handle resource. If the user has not already opened the file handle for writing ahead of time, they could easily just inline the `fopen` within the `toFile()` stream method call.

Smoren replied:
OK. [Done](https://github.com/markrogoyski/itertools-php/pull/19/commits/b0b7be2cee9ca39a029f25fe2faf742091a2d396).

Smoren committed:
`Stream` namespace: methods working with files are tested now.

markrogoyski commented on src/File.php:
```diff
@@ -0,0 +1,61 @@
+
+        while (($line = \fgets($file)) !== false) {
+            yield $line;
+            ResourceHelper::checkIsValid($file);
```
Question: Why is this check required after each line as it was already determined to be a valid resource to get into the while loop?

Smoren replied:
I agree with you. This is definitely an unnecessary check. Removed it.

Smoren committed:
`File` methods: checking resource in the reading process is removed.

Smoren commented:
Hi @markrogoyski,

I've added tests for all the methods.

But I have a problem with testing of `FileHelper`. Do you have any ideas about how to emulate non-writable file and directory?

Smoren committed:
`Stream`: file terminal operations take resources instead of file names as input params now. FileHelper and FileException are removed.

Smoren commented:
I think all the methods about working with files are added and tested:

* `File::readLines()`
* `File::readCsv()`
* `Stream::ofFileLines()`
* `Stream::ofCsvFile()`
* `Stream::toFile()`
* `Stream::toCsvFile()`

If you think so, I'll add the docs and you can merge this PR :)

UPD: branch is rebased from `develop`.

markrogoyski commented on src/Stream.php:
```diff
@@ -676,6 +712,61 @@ public function toArray(): array
+     *
+     * @return void
+     */
+    public function toFile($file, string $separator = "\n", string $prefix = '', string $suffix = ''): void
```
Question: Would it be more cross-platform compatible to use `\PHP_EOL` here instead of `\n` as the default separator here.

Also, I think `prefix` and `suffix` are misnamed here. I think what you want to call these are `header` and `footer`. And based on those names, I think a reasonable person would presume that the newline separator will be in between the header and lines, and the lines and the footer.

Smoren replied:
Done.

markrogoyski commented on src/Stream.php:
```diff
@@ -676,6 +712,61 @@ public function toArray(): array
+     *
+     * @return void
+     */
+    public function toCsvFile($file, string $separator = ',', string $enclosure = '"', string $escape = '\\'): void
```
If `toFile` has an optional _header_ text, then I think it would be reasonable to expect `toCsvFile` to also have it. I think it is a common case where you will have a data structure of all the rows of the CSV but that data structure will not contain the header row. And to make a CSV make sense in Excel or Google Sheets, you usually are going to want a header row defining the columns. Since the separator etc. is passed in, then it probably makes sense for the optional `headers` to also be an array. Thoughts?

Smoren replied:
Done.

Smoren committed:
1. `Stream::toFile()`: params renamed (`$prefix` => `$header`, `$suffix` => `$footer`).
2. `Stream::toFile()`: default value of param `$separator` was changed to `\PHP_EOL`.
3. `Stream::toCsvFile()`: param `array $header` was added.
4. New test cases added for `Stream::toCsvFile()`.

Smoren commented:
Hi, @markrogoyski,

All your suggestions were applied.

Summary:
1. `Stream::toFile()`: params renamed (`$prefix` => `$header`, `$suffix` => `$footer`).
2. `Stream::toFile()`: default value of param `$separator` was changed to `\PHP_EOL`.
3. `Stream::toCsvFile()`: param `array $header` was added.
4. New test cases added for `Stream::toCsvFile()`.

markrogoyski commented on src/Stream.php:
```diff
@@ -676,6 +712,71 @@ public function toArray(): array
+
+        $firstIteration = true;
+
+        fputs($file, $header);
+
+        foreach ($this->iterable as $line) {
+            if ($firstIteration) {
+                $firstIteration = false;
+            } else {
+                $line = $separator . \strval($line);
+            }
+
+            fputs($file, \strval($line));
+        }
+
+        fputs($file, $footer);
```
The newline separator would have to be added after the header if one is provided, and before the footer if one is provided, otherwise it isn't functioning as a header and footer. So I think you'd need some if statement prior to each to determine if you need to write them and put in the separator. Since empty line would be a valid header (and footer), perhaps the default value for header/footer should be null, then the if statement is if it is not null, add the header + newline separator, and similarly for the footer.

For example, if I do something like this:
```php
$header = '<ul>';
$footer = '</ul>';

Stream::of(['item1', 'item2', 'item3'])
    ->map(fn ($item) => "  <li>$item</li>")
    ->toFile($fh, \PHP_EOL, $header, $footer);
```

I'd expect the result to look like this:
```html
<ul>
  <li>item1</li>
  <li>item2</li>
  <li>item3</li>
</ul>
```

Smoren replied:
Done.

markrogoyski commented:
Hi @Smoren,

Looks good. Just one last comment about the header and footer.

Thanks!
Mark

Smoren committed:
`Stream::toFile()`: separators for header and footer added.

Smoren commented:
Hi @markrogoyski,

`Stream::toFile()`: separators for header and footer added.

Smoren commented:
Thank you Mark for merging this PR!
